cluster-operator/src/main/java/io/strimzi/operator/cluster/operator/assembly/KafkaConnectAssemblyOperator.java

```diff
@@ -359,4 +365,26 @@ Future<Integer> generateAuthHash(String namespace, KafkaConnectSpec kafkaConnect
     private boolean isPaused(KafkaConnectorStatus status) {
         return status != null && status.getConditions().stream().anyMatch(condition -> "ReconciliationPaused".equals(condition.getType()));
     }
+
+    /**
+     * Create Kubernetes watch.
+     *
+     * @param namespace Namespace where to watch.
+     * @param onClose Callback called when the watch is closed.
+     *
+     * @return A future which completes when the watcher has been created.
+     */
+    public Future<Watch> createWatch(String namespace, Consumer<WatcherException> onClose) {
+        watcher = new OperatorWatcher<KafkaConnect>(this, namespace, onClose);
+        return VertxUtil.async(vertx, () -> resourceOperator.watch(namespace, selector(), watcher));
```

Comment:
Why is this here? It replaces the existing method in the upper classes without any obvious reason.

Reply:
It is written this way to extract the watcher used to create the watch. The watcher is extracted for testing purposes, so we can see how the operator behaves with a WatcherException when we run onClose(e) on the watcher

Reply:
And how are the other watches tested without storing the watch? Maybe if they don't have a test, you do not need one either. Or you should add it for all of them (but I guess simulating the watch closing with some exception to test it is quite hard so that is why it has no test).

Reply:
There are not any tests for the other operators but I was thinking of adding one for them. Let me take a closer look into this, if I can align the onCreate() method of the KafkaRebalanceAssemblyOperator and the KafkaConnectorAssemblyOperator to the parent class, we could have one test that covers all operators

Reply:
Maybe the tests are not there because it is hard to test it. To be honest, I'm not sure what exactly the tests you have test. You don't really trigger a real watch exception. You just call the onClose handler. But you do not have control over the original watch nor over the new one. So I think your test does not prove anything really. The only way to really test it might be to mock the socket or mock the actual watch code inside Fabric8.

Reply:
But from the partial stack trace and report from the user's cluster in the issue we know that onClose() is called with an exception, more specifically a `KubernetesClientException` and that the watch is not recreated. If we know this, isn't the test enough to show how the the operator recovers from having its onClose() method closed by an exception like this is, even if we are throwing it ourselves in the test?

Reply:
Maybe it would make more sense if we threw a `KubernetesClientException` instead of the `WatcherException` to be more closely represent the issue?

Reply:
It is different whether you call the onClose manually or whether the watch calls it. For example in what the state of the original watch is.

Reply:
Removed